### ISSUE TITLE
Allow excluding thirdPartyCompatibility from GMM

### ISSUE THREAD

**Issue by samypr100** [gradle/gradle#26144]
### Expected Behavior

Per the GMM spec, I would like to have the ability to exclude `thirdPartyCompatibility` from the generated `module.json` since it's optional.

In summary, I would like to have the ability to do at least do the following:

```groovy
tasks.withType(GenerateModuleMetadata) {
    thirdPartyCompatibility = false
}
```

### Current Behavior (optional)

Currently when working with JavaFX projects, either using the JavaFX plugin or manually specifying the JavaFX dependencies, we get a `module.json` that looks like below:

```json
...
{
  "group": "org.openjfx",
  "module": "javafx-base",
  "version": {
    "requires": "20.0.2"
  },
  "thirdPartyCompatibility": {
    "artifactSelector": {
      "name": "javafx-base",
      "type": "jar",
      "extension": "jar",
      "classifier": "linux"
    }
  }
}
...
```

This can cause issues to downstream consumers such as this JLink error `Error: Two versions of module javafx.base found in /path/to/build/jlinkbase/jlinkjars (javafx-base-20.0.2-mac-aarch64.jar and javafx-base-20.0.2-linux.jar)` since the published .module file will have the `classifier` entry.

### Context

This is a follow up per @Vampire recommendation based on a [Slack Conversation](https://gradle-community.slack.com/archives/CAHSN3LDN/p1692137622964049).

In summary, I would like to have the ability to do at least do the following:

```groovy
tasks.withType(GenerateModuleMetadata) {
    thirdPartyCompatibility = false
}
```

To explain, lets consider the following `build.gradle` file for a typical OSS JavaFX library package:

```groovy
plugins {
    id 'java'
    id 'java-library'
    id 'application'
    id 'maven-publish'
    id 'org.openjfx.javafxplugin' version '0.0.14'
}

group = 'org.example'
version = '1.0-SNAPSHOT'
description = 'An example JFX app'

repositories {
    mavenCentral()
}

application {
    mainModule = 'org.example'
    mainClass = 'org.example.Launcher'
}

java {
    sourceCompatibility = JavaVersion.VERSION_17
    targetCompatibility = JavaVersion.VERSION_17
    modularity.inferModulePath.set(true)
}

javafx {
    version = '20.0.2' 
    modules = ['javafx.controls', 'javafx.graphics']
}

publishing {
    publications {
        mavenCustom(MavenPublication) {
            artifactId = 'example'
            group = project.group
            version = project.version
            from components.java

            pom.withXml {
                def root = asNode()
                root.dependencies.'*'
                        .findAll() { it.groupId.text() == 'org.openjfx' }
                        .each { it.remove(it.classifier) }
            }

            pom {
                name = 'Example'
                description = project.description
            }
        }
    }

}
```

When publishing this package, it yields the proper `pom.xml` as its common practice to remove the `classifier` from the JavaFX dependencies for consumers, but the `module.json` is currently a bit problematic.

The `pom.xml`:

```xml
...
<dependency>
  <groupId>org.openjfx</groupId>
  <artifactId>javafx-base</artifactId>
  <version>20.0.2</version>
  <scope>runtime</scope>
</dependency>
...
```

The `module.json`:

```json
...
{
  "group": "org.openjfx",
  "module": "javafx-base",
  "version": {
    "requires": "20.0.2"
  },
  "thirdPartyCompatibility": {
    "artifactSelector": {
      "name": "javafx-base",
      "type": "jar",
      "extension": "jar",
      "classifier": "linux"
    }
  }
}
...
```

This can cause issues to downstream consumers such as this JLink error `Error: Two versions of module javafx.base found in /path/to/build/jlinkbase/jlinkjars (javafx-base-20.0.2-mac-aarch64.jar and javafx-base-20.0.2-linux.jar)` since the published .module file will have the `classifier` entry.

I'm looking for an option that I can prevent this from happening and yield the following `module.json`:

```json
...
{
  "group": "org.openjfx",
  "module": "javafx-base",
  "version": {
    "requires": "20.0.2"
  }
}
...
```

which behaves as expected to downstream consumers when we publish our JavaFX libraries.

Currently, I'm resorting to the following code to modify the `module.json` before publishing:

```groovy
tasks.withType(GenerateModuleMetadata).configureEach {
    doLast { _ ->
        def gmmFile = it.outputFile.get().asFile
        def inJson = new JsonSlurper().parseText(gmmFile.text)
        def filteredVariant = inJson.variants.findAll { it.name == configurations.runtimeElements.name }
        // remove "thirdPartyCompatibility" from GMM
        filteredVariant.dependencies.first().each {
            if (it.group == 'org.openjfx') {
                it.remove('thirdPartyCompatibility')
            }
        }
        def outJson = JsonOutput.toJson(inJson)
        gmmFile.write(JsonOutput.prettyPrint(outJson))
    }
}
```

Here's an example from a published library before and after the fix above:

* Before: https://repo1.maven.org/maven2/eu/hansolo/fx/charts/17.1.47/charts-17.1.47.module
* After: https://repo1.maven.org/maven2/eu/hansolo/fx/charts/17.1.51/charts-17.1.51.module

**Comment by jjohannes** [comment]
I assume you get the "thirdPartyCompatibility" part in the Metadata, because you define the dependencies with `classifier`.

You could get rid of that by **not** defining a classifier in the dependency and instead:
- Define a component metadata rule that makes all the "Variants" of JavaFX known as proper Gradle Variants
- Select Variants in you own build based on attributes on the classpath(s)

I think JavaFX should publish Gradle Metadata with the corresponding variant information. And the JavaFX plugin should be updated to use that (or maybe it won't be needed at all).

Here is the rule that you could use: https://github.com/gradlex-org/java-ecosystem-capabilities/issues/57

Then don't define classified dependencies, but request attributes on the classpath configurations - e.g.:

``` 
sourceSets.all {
    configurations.getByName(compileClasspathConfigurationName) {
        attributes {
            attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, objects.named(LINUX))
            attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, objects.named("x86-64"))
        }
    }
    configurations.getByName(runtimeClasspathConfigurationName) {
        attributes {
            attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, objects.named(LINUX))
            attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, objects.named("x86-64"))
        }
    }
}
```

(and do not use the Java FX plugin)

**Comment by samypr100** [comment]
@jjohannes Thanks for that alternative approach. I had looked into that approach last week, but I was trying to find a meet-in-the-middle approach where we can help JavaFX library maintainers transition to (this issue is for the benefit of JavaFX library maintainers) without too much tech debt of having to maintain a running list of all JavaFX classifiers in their repos.

Some maintainers just [don't publish GMM](https://github.com/controlsfx/controlsfx/blob/master/build.gradle#L73) at all.
Another [alternative I saw](https://github.com/FXMisc/RichTextFX/blob/master/richtextfx/build.gradle) a different JavaFX library maintainer doing is below, which is sort off undesired since it removes stuff entirely from the GMM.
```groovy
   // Prevent pom & module.json from containing JavaFX runtime dependencies
    if ( taskNames.stream().anyMatch( t -> t.startsWith('pub') ) ) {
        configuration = 'compileOnly'
    }
```

I'm not sure if upstream JavaFX will be publishing a GMM anytime soon. I realize that's likely the best preferred solution, but given my uncertainty here if that will ever happen I opened this issue. I thought it was a fair proposal given that since `thirdPartyCompatibility` is optional per the spec that there should be a built-in way to exclude it in gradle.

I'm also not sure if it will be a straightforward solution to have every JavaFX library maintainer adopt what's proposed in https://github.com/gradlex-org/java-ecosystem-capabilities/issues/57 and drop the JavaFX plugin since the plugin is what's in the getting started docs for JavaFX, see https://openjfx.io/openjfx-docs/#gradle, so it's almost a JavaFX mantra at this point to use the JavaFX plugins 😆.

Having said that, I still would think this new config setting could help ease the burden on other similar scenarios, obviously the most notable being JavaFX here. Another good meet in the middle approach could be updating the JavaFX gradle plugin, but I'm unsure what the approach would be there without impacting consumers. Thoughts?

**Comment by Vampire** [comment]
I would definitely welcome if you fix that JavaFX mess @jjohannes :-D

**Comment by jjohannes** [comment]
I took a look at the plugin. It's not that much code actually. So well... I ended up with a full implementation for moving it to variant support and removing the "hard link" to `org.javamodularity.moduleplugin`. Lets see....

https://github.com/openjfx/javafx-gradle-plugin/pull/151

**Comment by ljacomet** [comment]
When the task to generate GMM was designed, it was a conscious decision to not allow users to easily mess with the generated content.

Given the conversation and the fact that there is a better path for JavaFX, I will close this as there is no plan to provide such an escape hatch.

**Comment by samypr100** [comment]
@ljacomet Thanks, that's alright to close as my main concern when I opened the issue was with JavaFX. I kept it open because its likely other oss libraries might suffer from the same situation even outside the JavaFX context. 

I still think something like this might still be useful in the long run and I don't see it as an escape hatch since per spec thirdPartyCompatibility is optional to begin with, but I understand that from Gradle's point of view and design choice it's something that is low priority.

**Comment by Vampire** [comment]
@samypr100 you are looking at this from a slightly wrong angle.
That it is optional in the spec means it is optional in the file, as it can be absent if not needed, i.e. no `classifier` is used.
But it does not mean that a user of Gradle should be able to leave it out.
The sense is, "configure your project correctly, then the correct GMM file is produced".
This has various advantages, for example also the POM is then most likely correct, and it also works in other situation like if you for example consume the build in a composite build.
If you would do the latter, it would behave differently than when consuming the published artifact as you manually manipulate the metadata for the publication, no matter whether you do it like in your `doLast` manipulation or via something like `pom.withXml`.
So I don't think this is "low priority", but a good and sane design decision, that is intended to force you fixing your build model instead of manipulating some file on publication.

The main problem with JavaFX is, that the JavaFX Gradle plugin in its current state does a hack-around to work-around the suboptimal JavaFX publishing. This hack-around causes the dependencies with classifiers to be used and thus "breaks" your model which is then persisted in the GMM. Your `doLast` hack, or a `skipThirdPartyCompatibility` escape hatch, is just another hack-around to revert the negative effect the other hack-around had on the produced GMM.

If the problem in the JavaFX publication is work-arounded properly, using component metadata rules to properly declare the variants like @jjohannes suggested is then less a hack-around, but a proper work-around that solves the problem in JavaFX publication without the negative effect on GMM.

If there are other "other oss libraries that suffer from the same situation even outside the JavaFX context", those situation should be reported separately in case there is no proper work-around, so that Gradle could provide a way to properly work-around the problem without needing to manually manipulate the produced GMM which then just shifts the negative impact to an even later time as I described above. :-)

**Comment by samypr100** [comment]
@Vampire Apologies, I do see both angles. I think I just keep going back to how easily this situation can happen without library publishers noticing.

Would issuing a warning if you're publishing a GMM that includes `thirdPartyCompatibility` be an acceptable path forward?

**Comment by Vampire** [comment]
In my opinion no, as it can be a perfectly fine situation, it just is not in the JavaFX case.
But that's on the Gradle folks to decide, not me. :-)

**Comment by samypr100** [comment]
Yes, I can see it being perfectly fine as well, but the lack of corresponding variant information can be a problem too 😂
